crates/net/network/src/transactions.rs

```diff
@@ -1074,116 +1279,355 @@ struct Peer {
+        let Some(inflight_count) = self.active_peers.peek(&peer_id) else { return true };
+        if *inflight_count < MAX_CONCURRENT_TX_REQUESTS_PER_PEER {
+            return true
+        }
+        false
+    }
+
+    /// Returns any idle peer for the given hash. Writes peer IDs of any ended sessions to buffer
+    /// passed as parameter.
+    fn get_idle_peer_for(
+        &self,
+        hash: TxHash,
+        ended_sessions_buf: &mut Vec<PeerId>,
+        is_session_active: impl Fn(PeerId) -> bool,
+    ) -> Option<PeerId> {
+        let (_, peers) = self.unknown_hashes.peek(&hash)?;
+
+        for &peer_id in peers.iter() {
+            if self.is_idle(peer_id) {
+                if is_session_active(peer_id) {
+                    return Some(peer_id)
+                } else {
+                    ended_sessions_buf.push(peer_id);
                 }
-                Err(_) => {
-                    // TODO: re-request missing hashes
-                    self.remove_inflight_hashes(&requested_hashes);
-                    // request channel closed/dropped
-                    Poll::Ready(FetchEvent::FetchError {
-                        peer_id,
-                        error: RequestError::ChannelClosed,
-                    })
+            }
+        }
+
+        None
+    }
+
+    /// Packages hashes for [`GetPooledTxRequest`] up to limit. Returns left over hashes.
+    fn pack_hashes(&mut self, hashes: &mut Vec<TxHash>, peer_id: PeerId) -> Vec<TxHash> {
+        let Some(hash) = hashes.first() else { return vec![] };
+
+        if self.eth68_meta.get(hash).is_some() {
+            return self.pack_hashes_eth68(hashes, peer_id)
+        }
+        self.pack_hashes_eth66(hashes, peer_id)
+    }
+
+    /// Packages hashes for [`GetPooledTxRequest`] up to limit as defined by protocol version 66.
+    /// If necessary, takes hashes from buffer for which peer is listed as fallback peer.
+    ///
+    /// Returns left over hashes.
+    fn pack_hashes_eth66(&mut self, hashes: &mut Vec<TxHash>, peer_id: PeerId) -> Vec<TxHash> {
+        if hashes.len() < GET_POOLED_TRANSACTION_SOFT_LIMIT_NUM_HASHES {
+            self.fill_request_for_peer(hashes, peer_id, None);
+            return vec![]
+        }
+        hashes.split_off(GET_POOLED_TRANSACTION_SOFT_LIMIT_NUM_HASHES)
+    }
+
+    /// Returns `Ok(true)` if hash is included in request. If there is still space in the
+    /// respective response but not enough for the tx of given hash, `Ok(false)` is returned.
+    /// Throws error if `MAX_FULL_TRANSACTIONS_PACKET_BYTE_SIZE` is reached.
+    fn include_eth68_hash(
+        &self,
+        acc_size_response: &mut usize,
+        eth68_hash: TxHash,
+    ) -> Option<bool> {
+        debug_assert!(
+            self.eth68_meta.peek(&eth68_hash).is_some(),
+            "broken invariant `eth68-hash` and `eth68-meta`"
+        );
+
+        if *acc_size_response == MAX_FULL_TRANSACTIONS_PACKET_SIZE {
+            return None;
+        }
```

Comment:
I think this is probably very unlikely, I see why this check is useful as an early exit condition when we fill the request, but we could also move this below the if statement in L1413 ? This way we optimistically try to include the hash and only check if the request already completely full.

Though since we already compare `next_acc_size <= MAX_FULL_TRANSACTIONS_PACKET_SIZE` we could handle this case with `match cmp() {}` directly, we'd need a different return type that can express this case, so perhaps not worth the effort for a single == check

Reply:
I see your point. I wanted to consolidate the eht68 size logic into one call in `fill_request_for_peer`, and that check isn't expensive in `include_eth68_hash`. since I haven't put the size logic for eth66 in it's own method however, checking if the eth68 size limit is reached in `fill_request_for_peer` directly, isn't breaking any pattern.
```rust
} else if hashes.len() >= GET_POOLED_TRANSACTION_SOFT_LIMIT_NUM_HASHES {
        break
}
```